src/analysis/sample.jl

```diff
@@ -14,25 +14,193 @@
+        while iter < iter_target
+            iter += 1
+
+            center .= sum(sample, dims = 2)[:] ./ n
+
+            for i = 1:n
+                dir = center .- sample[:, i]
+
+                # iteratively collect the maximum and minimum possible multiple
+                # of `dir` added to the current point
+                run_range = (F(-Inf), F(Inf))
+                for j = 1:d
+                    run_range = update_range(
+                        run_range,
+                        sample[j, i],
+                        dir[j],
+                        variable_lower_bounds[j],
+                        variable_upper_bounds[j],
+                        epsilon,
+                    )
+                end
+
+                # do the same for coupling constraints
+                dc = coupling * dir
+                pc = coupling * sample[:, i]
+                for j = 1:nc
+                    run_range = update_range(
+                        run_range,
+                        pc[j],
+                        dc[j],
+                        lower_bounds[j],
+                        upper_bounds[j],
+                        epsilon,
+                    )
+                end
+
+                # generate a point in the viable run range and update it
+                lambda = run_range[1] + rand(generator) * (run_range[2] - run_range[1])
+                isfinite(lambda) || continue # bail out to avoid divergence
+                sample[:, i] .+= lambda .* dir
+            end
+        end
+        result[:, n*(iter_idx-1)+1:iter_idx*n] .= sample
+    end
+
+    return collect(result')
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+TODO
+"""
+function sample_constraint_variables(
+    sampler::Function,
+    constraints::C.ConstraintTree;
+    start_variables::Matrix{Float64},
+    seed::UInt,
     workers = D.workers(),
-) end
+    n_chains,
+    kwargs...,
+)
+    # preallocate a bit of stuff
+    d = size(start_variables, 2)
+    @assert C.var_count(constraints) <= d
+    lbs = fill(-Inf, d)
+    ubs = fill(Inf, d)
+    cs = C.Constraint[]
+
+    # gather constraints information
+    tick(v::C.LinearValue, b::C.EqualTo) = tick(v, C.Between(b.equal_to, b.equal_to))
+    tick(v::C.LinearValue, b::C.Between) =
+        if length(v.idxs) == 1 && first(v.idxs) > 0
+            idx = first(v.idxs)
+            lbs[idx] = max(lbs[idx], b.lower)
+            ubs[idx] = min(ubs[idx], b.upper)
+        else
+            push!(cs, C.Constraint(v, b))
+        end
+    tick(_, ::Nothing) = nothing
+    tick(v, b) = throw(
+        DomainError(
+            (v, b),
+            "sampling from unsupported constraint type: $(typeof(v)), $(typeof(b))",
+        ),
+    )
+
+    C.traverse(constraints) do c
+        tick(c.value, c.bound)
+    end
+
+    # convert constraints to a matrix
+    coupling = vcat(
+        transpose.(SparseArrays.sparsevec(c.value.idxs, c.value.weights, d) for c in cs)...,
+    )
```

Comment:
yeah well, I want to avoid precisely that -- any "simple conversion" is biased by design (there are too many interpretations and neither of them is uniquely "natural").